packages/versioning/test/test-incomaptible-versioning.ts

```diff
@@ -0,0 +1,354 @@
+    });
+
+    it("emit diagnostic when op was removed after return type", async () => {
+      const diagnostics = await runner.diagnose(`
+        @removed("2")
+        model Bar {
+          @removed("3")
+          foo: string;
+        }
+      `);
+      expectDiagnostics(diagnostics, {
+        code: "@cadl-lang/versioning/incompatible-versioned-reference",
+        message:
+          "'TestService.Bar' was removed on version '2' but contains type 'TestService.Bar.foo' removed in version '2'.",
+      });
+    });
+
+    it("succeed if version are compatible", async () => {
+      const diagnostics = await runner.diagnose(`
+        @added("2")
+        @removed("4")
+        model Bar {
+          @added("2")
+          @removed("3")
+          foo: string;
+        }
+      `);
+      expectDiagnosticEmpty(diagnostics);
+    });
+  });
+
+  describe("interface operations", () => {
+    it("succeed when unversioned interface has versioned operation", async () => {
+      const diagnostics = await runner.diagnose(`
+        interface Bar {
+          @added("2")
+          foo(): string;
+        }
+      `);
+      expectDiagnosticEmpty(diagnostics);
+    });
+
+    it("emit diagnostic when model property was added before model itself", async () => {
+      const diagnostics = await runner.diagnose(`
+        @added("3")
+        interface Bar {
+          @added("2")
+          foo(): string;
+        }
+      `);
+      expectDiagnostics(diagnostics, {
+        code: "@cadl-lang/versioning/incompatible-versioned-reference",
+        message:
+          "'TestService.Bar' was added on version '3' but contains type 'TestService.foo' added in version '2'.",
+      });
+    });
+
+    it("emit diagnostic when op was removed after return type", async () => {
+      const diagnostics = await runner.diagnose(`
+        @removed("2")
+        interface Bar {
+          @removed("3")
+          foo(): string;
+        }
+      `);
+      expectDiagnostics(diagnostics, {
+        code: "@cadl-lang/versioning/incompatible-versioned-reference",
+        message:
+          "'TestService.Bar' was removed on version '2' but contains type 'TestService.foo' removed in version '2'.",
+      });
+    });
+
+    it("succeed if version are compatible", async () => {
+      const diagnostics = await runner.diagnose(`
+        @added("2")
+        @removed("4")
+        interface Bar {
+          @added("2")
+          @removed("3")
+          foo(): string;
+        }
+      `);
+      expectDiagnosticEmpty(diagnostics);
+    });
+  });
+
+  describe("with versioned dependencies", () => {
+    let runner: BasicTestRunner;
+
+    beforeEach(async () => {
+      const host = await createVersioningTestHost();
+      runner = createTestWrapper(host, (code) => code);
+    });
+
+    it("emit diagnostic when referencing incompatible version via version dependency", async () => {
+      // Here Foo was added in v2 which makes it only available in 1 & 2.
```

Comment:
Here is where it would be good to have a test over implicit addedOn - if a namespace's lowest version depends on version "12" of versionedLib, then *every* type in the namespace should be able to use VersionedLib.Foo